src/gui/qgsmaptoolcapture.cpp

```diff
@@ -289,10 +291,40 @@ bool QgsMapToolCapture::tracingAddVertex( const QgsPointXY &point )
     mSnappingMatches.append( QgsPointLocator::Match() );
   }
 
+  // If the layer supports curves, we de-approximate curves
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() );
+  if ( vlayer->dataProvider()->capabilities().testFlag( QgsVectorDataProvider::Capability::CircularGeometries ) )
+  {
+    // We convert the capture curve to OGR geometry
+    OGRGeometry *ogrCaptureCurve;
+    OGRGeometryFactory::createFromWkt( mCaptureCurve.asWkt().toLocal8Bit().data(), nullptr, &ogrCaptureCurve );
+
+    // We de-approximate the curves using OGR (the geometry must be forced to linear first, as a mCaptureCurve is a CompoundCurve)
+    OGRGeometry *ogrLinestring = ogrCaptureCurve->getLinearGeometry(); // it must be forced to linear, as captuecurve is a compound curve
+    OGRGeometry *ogrRecurved = ogrLinestring->getCurveGeometry();
```

Comment:
@rouault is OGR_G_GetCurveGeometry the same algorithm as Postgis' ST_LineToCurve ? Or is it a different algorithm? I've previously tried porting Postgis' ST_LineToCurve to a native QGIS function (See https://github.com/nyalldawson/QGIS/commit/ac8061ef6e0556fb3befbfca95d13f15c4d2e2f2), but found a few issues with the results from that function in Postgis (and never got the time to fix them unfortunately)

Reply:
I noticed some glitches on certain geometries : in certain cases, some arcs get completely removed... not sure why... is this the type or issues you encountered ?

As said above (https://github.com/qgis/QGIS/pull/35875#discussion_r411488123) having this available in QGIS would definitely make sense, it would be awesome if we could solve the issues you were mentionning.

Reply:
> OGR_G_GetCurveGeometry the same algorithm as Postgis' ST_LineToCurve ?

Much likely not. My own creation. Not something fully bullet proof. Definitely not as I discussed with @m-kuhn  some time ago
```
Some years ago I did for fun a hack. When using OGR geometry methods that ultimately rely on GEOS methods, I first linearize geometries, but hide in the LSB of some X&Y coordinates
a hidden information (see the ADD_INTERMEDIATE_POINT=STEALTH
option of https://gdal.org/api/ogrgeometry_cpp.html#_CPPv4N18OGRGeometryFactory17curveToLineStringEdddddddddidPPCKc ),
so that when GEOS is done, I can re-use to, possibly, rebuild circular portions that
weren't affected by the geometry operation. This is of course a super gross hack, not robust at all...

Little demo though on a simple case with 2 circles overlapping:

>>> from osgeo import ogr
>>> g1 = ogr.CreateGeometryFromWkt('CurvePolygon(CircularString(0 0,1 1,0 0))')
>>> g2 = ogr.CreateGeometryFromWkt('CurvePolygon(CircularString(0.5 0,1.5 1,0.5 0))')
>>> print(g1.Union(g2))
CURVEPOLYGON (COMPOUNDCURVE ((0.749706234848466 -0.16111841890272,0.721987566409211 -0.171357967376366),CIRCULARSTRING (0.721987566409211 -0.171357967376366,-0.206869769243333 0.481693462099955,0.733542564392451 1.16742630351001),(0.733542564392451 1.16742630351001,0.749699897508666 1.16113860048096,0.754972022079414 1.16329577869618),CIRCULARSTRING (0.754972022079414 1.16329577869618,1.70581670320598 0.542694044965174,0.789633573195289 -0.175089598848494),(0.789633573195289 -0.175089598848494,0.749706234848466 -0.16111841890272)))

But I wouldn't recommend relying on that. Due to the linearization, and how reconstruction
is done, there are certainly inaccuracies that would make it unfit when topological
correctness is needed. Like in the above example where there are 2 small segments, that
shouldn't be there: the union should be just 2 circular strings.
```
Code at https://github.com/OSGeo/gdal/blob/master/gdal/ogr/ogrgeometryfactory.cpp#L5858 , and core of it at https://github.com/OSGeo/gdal/blob/master/gdal/ogr/ogrgeometryfactory.cpp#L5289